### ISSUE TITLE
FreeBSD Port Team: Volunteers needed

### ISSUE THREAD

**Issue by janhenke** [dotnet/coreclr#455]
Following the discussion on the chat, I am looking for volunteers to start a FreeBSD port team.

The purpose is to have a group of people working together on making coreclr working well on FreeBSD. This includes having more people with FreeBSD experience for reviewing patches and keeping an eye on PRs.

This issue is intended to discuss this idea and as a means to get into contact with each other. Also technical discussions about porting to FreeBSD are possible.

**Comment by xied75** [comment]
My experience with FreeBSD is making VM template for CloudStack, if this can qualify I would like to join the team and picking up this brilliant OS on the way. :)

**Comment by MattWhilden** [comment]
it's likely that you're already aware of this fork but if not it's worth checking out: https://github.com/ajensenwaud/coreclr

**Comment by MattWhilden** [comment]
Ugh. Was too fast on the comment button. Meant to attempt to summon @ajensenwaud to the thread.

**Comment by richlander** [comment]
To provide some more context ... to make this project awesome, we'd want a core group of folks that felt ownership and accountability (in a good way) to make the project happen. In turn, the .NET Core team would be more compelled to invest in CI infrastructure for FreeBSD as a first class citizen.

I like the idea of "port team". Some of very best people the CLR team has ever had were on the "X64 port team", 10+ years ago. They simply referred to themselves as "the port team". Not the type of folks you'd want to meet in (technical) dark allies ;)

**Comment by kangaroo** [comment]
While I don't have the time to contribute significant resources to this, given the familiarities with OS X, I'm happy to support with code review, direction and suggestions.  Feel free to @kangaroo me on issues.

**Comment by ghuntley** [comment]
Long time FreeBSD user. I'd be keen to help out from an administrative/marketing/documentation role where needed as my C++ foo is weak and this is a good chance as any to get better :-)

**Comment by ghuntley** [comment]
Summoning @josteink , @Aesthetikx , @ajensenwaud and @kangaroo who the lads behind https://github.com/ajensenwaud/coreclr

**Comment by richlander** [comment]
These PRs are relevant to this discussion:

https://github.com/dotnet/coreclr/pull/453
https://github.com/dotnet/coreclr/pull/470

Nice to see that they have been merged.

**Comment by josteink** [comment]
I mostly tried to help out in @ajensenwaud's repo as a sort of "challenge accepted" short-term project.

As I don't have any up-to-date FreeBSD production-systems, I doubt I'm a good candidate for a longer term FreeBSD "port team". I do agree about the name though. It has a nice ring to it. Go for it!

That said, I appreciate being summoned and namedropped in `dotnet/coreclr`. That's almost going on my resume ;)

**Comment by ajensenwaud** [comment]
Hi all

Thanks Jostein, appreciate your help so far!

I am very keen to still assist with the porting and can also assist with test infrastructure, as required. How many are in for this?

Anders

**Comment by qbit** [comment]
Maybe we should make a wiki with status of each port, next steps and hurdles? OpenBSD is stuck until libunwind is ready (unless coreclr people can confirm that isn't a hard dep?).

**Comment by janvorli** [comment]
You can start porting without the libunwind. It is required for exception handling and GC. So as long as an application doesn't throw an exception or initiate a GC, everything should work without the libunwind.
That's the way we were bringing up coreclr on Linux, the exception and GC support came quite recently.

**Comment by qbit** [comment]
:+1: - I will keep hacking away then!

**Comment by janhenke** [comment]
Hi folks,

I did not answer earlier as I want to know how big the feedback is. So far I am happy to see quite some people eager to join this project.

One question we should answer though is whether we want separate *BSD teams (one for FreeBSD, one for OpenBSD, one for NetBSD) or one big BSD Port Team.
My experience is limited to FreeBSD so far, which by my understanding is also the most often used of the different *BSDs. But from what I read there seems quite some differences e.g. between OpenBSD and FreeBSD. So I have the tendency to limit the scope of the "FreeBSD Port Team" to FreeBSD only for the time being.

As @richlander proposed I would start a PR with a new Wiki page, describing the intention of the team, current members, agenda and way to participate. Currently I see as members:
 * @janhenke
 * @xied75 
 * @kangaroo 
 * @ghuntley 
 * @josteink 
 * @ajensenwaud 
Also @qbit for OpenBSD.
As a rought list of people interesting. Would be nice to get an answer to the FreeBSD <-> OpenBSD <-> NetBSD question and a ack/nak to team memebrship from everyone in the list. Then I can prepare the wiki page based on that outcome. Best reply until Friday noon UTC. :)

**Comment by ghuntley** [comment]
Alright, have created an Azure virtual machine (Basic_A2 - 2 cores, 3.5GB memory) running FreeBSD 10.1-RELEASE which @josteink (and others) can call home for the foreseeable future whilst this issue is open.

Post a public ssh-key here or email me (ghuntley@ghuntley.com) and I will create you an account on ```portteam.cloudapp.net``` that has sudo access.

**Comment by josteink** [comment]
I'm pretty busy these days, so I'definitely m not making any promises.

Anyway, I appreciate @ghuntley's initiative, and to show that, here's an SSH-key which means that in *theory* I can contribute should I ever get the time:

    ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDMaGl5KG5P6fogg5B5tY9K26X2rx0d7MZg7x28TYxICZGK1aHCODylNy3WfnBFvQ5hRsrf0kCT2u36JUmgJzXTBx6Vc3AchrEOMJLRl3KvJfsbpANrymkW666sDgQW6PyqXSZ8f17vREGZrvFa41/dwAxMvxlQELwS8HRgNzY8GoZBP9BTm6qqZFozBf25nZLN/p3/X3w3GxW4USJutnH5T2oBJc3TqVnBFSgrsnHMMlmV+Xj0EMBy4di+JoNBm78hCDwt1QVMJ+NDOH224LuWgTlhOIRRDZfNee3qGHOB1llI78BZkc4x/9SJR8Fiv13pUINHc1zfLBzmB1PPLztLS2j8kOTh/ymr3Gc5+8iGgPX9ZideNLoTjD9necwQ+GhPdys/DkU8idYkCJnVPv6il+OOwk55xopyyfs11660xlAxJKdNy8D3lbSBSQQCGMRKAXhXxUsQr1R64fScZ0NejmpXwYR+9vybymKbiMWizxYQ8ap/KxOAs3znPWYpu2mjldkHqKd0Wc5z4/pgZS/m2KVHr95oy1Lx/O77x6QSy9h7Wu1xIcayV5R+wIvfz6BhQ7Hpr2Lu2MHpNQr2T4UzIodfSJ/LWp/xo37Fl7Bg9SntEC4NepSW9FFSJ0q48/CfcTJYMvLTvqGYn85Y8PnlrKsr295RiRyMiD3Ys4RNtw== jostein@irssi

Anyone else?

**Comment by josteink** [comment]
> But from what I read there seems quite some differences e.g. between OpenBSD and FreeBSD.

From what I can tell, coreclr is pretty well abstracted and most (everything?) platform-specific goes through the Platform Abstraction Layer (PAL). If that builds cleanly, the build as a whole goes quite a bit further without too much trouble.

The PAL seems focused on underlying implementations rather than platform (see [context.cpp](https://github.com/dotnet/coreclr/blob/master/src/pal/src/arch/i386/context.cpp)): That checks if this architecture has BSD regs, PT regs, GREG-sets, what register-mapping etc.

Since I'm not sure if this is a problem at all, the best plan I can think of is to closely collaborate with the OpenBSD team (@qbit. anyone else?) and make sure the commited changes are mutually compatible.

If we later discover there's any substantial difference between FreeBSD and OpenBSD (or for instance Darwin), we might need more differentiation between the platforms than just saying "it's BSD", but let's save that for later, if for nothing else, just to get going

>  Alright, have created an Azure virtual machine ... Treat it with respect but don't be afraid to break it as it can be recreated as needed with a simple shell script.

If there's any global prerequisites/requirements will they be embedded in this script? According to the best of my knowledge @kangaroo's changes to support Darwin means clang3.7 (which we will be cherry picking some) will be required to build this. Clang3.7 is not installed by default in FreeBSD 10.1, nor is it the default c-compiler.

It would be nice to have stuff like this (and other things we may add incrementally) available to everyone. Basically a public build-script for the machine. Sounds doable?

Which leads us to the next issue. Where do we work?

Since I doubt we'll be given commit rights into coreclr directly, should we have a place of sorts to work where we can commit incremental improvements? If so @ajensenwaud already has [something going](https://github.com/ajensenwaud/coreclr).

Should the work be based on that, or should we "start over", based on the current coreclr and just port the fixes we have so far?

**Comment by janhenke** [comment]
Thanks @ghuntley for the sponsoring of a porting machine. My SSH pub key is
```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAAEAQDEhh5lz6n4LN6S9McCGCrDrr+rZNtY4PqMc0aRK2hND14UrZo0tfO9nPY9rZMhnJ8GUdOWbcX79l1IkEMZTI62rPP4TL65hhLDQBD6WeXy4ib8VFFwOjjmHt8BOctAAsaWQOich3kLcigMCyvn7HA4YXTvqwfY2M5AK6yCrBcWdfAcAaSv45uX7XCwg4xWd/xrwV6aE+rFPsbZCEcb0s/5MQ+/exILHuLMXxgJfUC7a+3HkJLyUZWWLt1JfVAz67/xPe5Z2P+eYF98Xq21T2Ptl2p8HFDgV6OpqfOkVYJrjriUxd4OABFPzsh63qaOmtvTT5VAMED/Z7+5nt7TY+lCLpYTt8ncAoIXoqbu0St1E121ZChsw9Y2Ltuj3Z9FSP69x9jogREJnC5TyTZOxNjoIVKD4o6zFYlls9fqqXC8tynJT6vrn0GxfC9ew4lD/Nin+97vwbZAJZVHqLBNpcDLCMIR3PTgc7TsUDrTUce6vsigczbp/69rJPpRqkPZcKPXT5kokW8z4gEE1/iICsZouoaVcZNCRiLt+PJO2nUlnYJQ5dIWOCJivxSAmZgk7RyHsc8zvLunrhVnse4tF+pTfa8pc5BJ23E2VIeVUbtwqrd2IwcuRoJy0E8yz/lJrZAVffGcXSOty6MM3fd68QeekbZpy/lT5FN73O6oU4TFGy+yVL368yFXXIjcfSQSG0sMaZFGXHdJwfJnJEdZgxuOCqaA3SooC5pnuC5pOF2Mt95RgahYPFo4GMU17AhpOSrJBz1zo94FCHcKUdsF4P9+GE6810q/uw8QwLOniw8IoBUP04ikbBSa95C2Qx6t1ygujrAxb+LJTtJ/X619t2CuCrHnMf2Vh+UNWtRbjuof5QWbO0GI97zt46vJifJC0w8fmuvOxP7bDdEITSWztPTgXftt0X7EVYXnf6dg2mGE/JWOZEfnspvbuYetkqCI+dqVWfuVodHGZ8H7nJr2XfEwxs+NLSOZkLu9v0EuJF/lviFbmpedBhCmfiqjo8POWU7jZ1MFSYGxqGsXHbL2b2/xr92aRqo3ZkX9NsqA4ofINePeGwuFMTpDMq0VWWW6V2Y5E3t8Noje8TRnN9uIZjLjLhBSVcwcmQivwF+GF6yi/osHnSHXYvajFQdkYuxRUlUeOh0MJIPExOs//jZ65srTL5yjEx4zkBn6Qsy4EPUGdchB8gPsZEtRt8zMGCPuoCL+3G4UOrjAxnLbbxb/PrIqvWKeFgvYnFMZSONA3BAYiAHKy67EqWq1yYttCyLQUT9R2R+wfMLlgCeWP9YGBDY8ArxUpPQzJZPJtnIo0ny+Sd0JFemql07jT7KTKlNL0vKaOpiGpBP11D7KHbKtIxI3 JanHenke_BSD_Port_Team
```

@josteink the major difference that came to my mind is the focus of OpenBSD on security. I read they introduced their own special APIs to make software more secure and less vulnerable to security related bugs. Admittedly though I do not know any details and just read that.

If it works from a compatibility side, it certainly would be nice to see all three BSDs supported. Though I think at the beginning FreeBSD should be the priority in case something does not work on all three the same way.
We could also organize it in a way that we all are the BSD Port Team and we assign people according to their preference to one or more "subteams", one for each BSD flavour.

Clang 3.7 is the current development IIRC. Clang 3.5 should be fine for the moment.

For the work, I was thinking about creating a Github "Organization", which could host our fork of the coreclr tree. @richlander how is your feeling about this? Also given that it would need a matching name with CoreCLR or dotnet in it. Maybe one member of the DotNet Foundation or Microsoft can set that up so it is guaranteed there also someone with administrative access at hand?

In any way I think it would be better to have central tree somewhere and not everybody having their private trees.

**Comment by josteink** [comment]
> Clang 3.7 is the current development IIRC. Clang 3.5 should be fine for the moment.

Fair enough. We'll stay on 3.5 unless it causes issues.

> I was thinking about creating a Github "Organization"

Seconding this. How about just "DotnetBSD"? What I'm not sure about is do we need central support for the DotNet Foundation or Microsoft to do this?

> In any way I think it would be better to have central tree somewhere and not everybody having their private trees.

Absolutely positively agreed.

**Comment by qbit** [comment]
Thanks @ghuntley ! 

    ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDNLvc26CLuM48MI0V6jQQVYZ5f3nMMPcX1iQPjfnHf9Wp9qlHaRUKYyveh4qcrfpVObvHfHbFZwyNQ+Ri7xfs0AFJLNiATKd50ditWT9FZ//xiGBuhpn2UwiYYTuCywkVuXbEdZXQesq9BrHhNn8ifmhPseKoMGOEil+eJGvxXnR3thiWGwuNSZ4D24brbVz3Hm/Fv9Gnbe3flui4jf73OxCyAT82NZ2FzbKCms5Q2QptwoAoDVTqgOar4lWX+vsZvs08EJ/HIRGTeHF/Yk9j8azJFHYxm6ao4eeQ9fHg5vXpsc/NX0VDm4ncfeuP/yVkxo7ZQC8yW+SkMlnj05CK9 qbit@slip.bold.daemon

> Since I'm not sure if this is a problem at all, the best plan I can think of is to closely collaborate with the OpenBSD team (@qbit. anyone else?) and make sure the commited changes are mutually compatible.

This works for me. Also trying to recruit other OpenBSD devs.

> the major difference that came to my mind is the focus of OpenBSD on security. 

This isn't so much of an issue - most of the changes have "standard" counterparts that give you stern warnings when using them. The bigger difference will be availability of ports / required things (libunwind..) / location of installed ports.

**Comment by richlander** [comment]
My morning now, so time to reply. 

I'd be happy to setup a coreclr-freebsd (or some similar name) repo as a fork of coreclr within the dotnet org and give the port team commit rights to it. To my mind, that provide the following:

- Unfettered development process (you control merging PRs).
- High discoverability of the repo (good to get more people involved).
- It's easy to setup the CLA and CI process on the repo. 
- .NET Core team has access to the repo should that be needed.

PRing your changes to upstream on a regular basis would still be a good idea, but I think that is your plan, anyway.

I haven't asked anyone else's opinion on this, like @jkotas or @martinwoodward, but this is my proposal, and I think its a good one.

**Comment by jkotas** [comment]
I do not think we need a fork for the FreeBSD port. I expect that the changes for the FreeBSD port are going to non-controversial, and they can be submitted to master as they are made.

**Comment by kangaroo** [comment]
I agree with @jkotas -- I expect the FreeBSD port will be far less invasive than the OS X one was.

**Comment by richlander** [comment]
What do you think about @janhenke?

**Comment by janhenke** [comment]
My idea was a bit different. I thought about having a Port Team Organization and repo primarily to directly work in it. So I would no longer have my private janhenke/coreclr tree, but there would some dotnetbsdteam/coreclr, where I can commit code to and open feature branches. In that way someone else in the team can work together with me on a feature without needing PRs for every commit.

PRs would only be used once we think a certain feature is ready so I can merged into the main sources.

So I do not want to have "parallel tree" here, but I rather thought a working place where we can work cooperatively in feature branches until they are ready to be merged. So by definition all changes this tree has from upstream is consider unstable until a PR is opened.

I also would not accept any PR into that tree, all PRs should go directly into dotnet/coreclr. Also the CI is only needed on the main dotnet repo, expanding the current bot simply with FreeBSD/OpenBSD/NetBSD instances.

**Comment by kangaroo** [comment]
I'm pretty strongly against this direction, as I can see it spiraling out of control quickly.  Having a separate tree per os?  Per platform?

Is there a compelling reason for the primary maintainer to not just have it on their personal fork, and accept PRs themselves and push to coreclr as needed?

**Comment by janhenke** [comment]
More centralization and overview what is being done. I do agree that it seems a matter of which kind of work flow we want to establish with this team.

If the majority does not like the idea, it is fine too. As I said I just want us to agree one common procedure to organize the work flow and also somehow make it possible to create a team where we can recognize each other and issues that affect us more easily. In the end it would also be good to have one easy way to summon the team as a whole to an issue or PR if the opinion of a BSD porter is needed.

**Comment by kangaroo** [comment]
I'm not sure why we're focusing on centralization outside the canonical repo.  MS has made it clear they'd be happy to upstream the patches, so lets just centralize here like for all other OSs.

In terms of having people recognize the issues, that seems like an easy candidate for a label?

**Comment by xied75** [comment]
Dear all,

Here is my ssh key:

```
ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAyVtbcfl/UqP8syjU0aEKRxr3+EOkmFWMZcR/xTSDo5unAqp3QX6G0qEp7upcQdvIrr7YUAkS3JupxRuaTQtSpb/qbP+Zy9bN2ktaylLg2LycyT5NKS2V5gDrO98ekM7/fxIICX+JShbT54IUReE7IwC5HlW0QNLOV2lkCt5BQoYKukqfs3t2J5OCuUuitmVugJ0QPtBhOVWdtvk48YiTlmS+NTWCoF/O9SxnctKTW4Fi90jIpfO7s/Onw/dD6GF+cxBykQRXiW6ukj4HY+oqzCkHvwzgfsS2NfnLZCE2q9lYvwLKTfOWca5C+fKf5/JCn7D8X9I+m7wxWeEwXk2KQQ== xied75@gmail.com
```
Please somebody kindly advise what shall I need to do/read to prepare for this project: e.g. read about FreeBSD, LLVM, Clang, or the source code, etc.

Thanks all.

**Comment by ajensenwaud** [comment]
Great idea with the VM on Azure!

Here is my ssh key:

ssh-dss
AAAAB3NzaC1kc3MAAACBAI0brhrRR9wMD9xm62O7MkQBpX2ASl0lGue4L+y2KFptZmfoioQt
ZjPoBgjhKaG64uPH6V3MLzHGJn2iiSHHWxyA+2WHK1K3rQt6mEiIiU4N3PmI140wKJpuoyoCnKwScTGC
P7VYT5nJH5yKTz/nRNmS1llFkfaSfSxEXhgT0MYHAAAAFQDG6uDfY9dUBFYiPPLtV/7yCG4+owAAAIEA
iYVmRHXgBE6jiQ7GogQ/3ZqdoSu6bufyIN1kG7p9xEgFHeH8V+wPczb/RmS47X39qNraI3RcZ+zIaSrF
s1M9uyhA1MIYGQ4P3kXS+H+4Wydq3SUCINfO7oGEgxsLCnDo6gdz1//HZqaWRiazM7mtpBzeO1rP0W5e
o5BJyW2L1EsAAACAKgo7Kz2EL993yXFJJGFNWomZiYviwMCxSGF/zrjF3vJt8QrUJyvBfDosD6HUcfok
qdUBd4ztOOPT6fHVE1Oa1lOCQe5731+sk1Zq5bA4BDnOmhfu/LZHAH6GctPOegb9U0QHcVrhzt6okMtB
8yiFrFiX/TiixXORHnQGt4HKQWs= aj@beastie.jensenwaud.com

**Comment by ajensenwaud** [comment]
On Fri, Mar 20, 2015 at 7:46 AM, Geoff Norton <notifications@github.com>
wrote:

> I'm not sure why we're focusing on centralization outside the canonical
> repo. MS has made it clear they'd be happy to upstream the patches, so lets
> just centralize here like for all other OSs.
>
> In terms of having people recognize the issues, that seems like an easy
> candidate for a label?
>

Agree, but unless all of us get commit access to coreclr, we are not going
to be able to consolidate our patches before submitting them to upstream?

For sake of separation, I suggest building out github.com/dotnetbsd  unless
we can all work out of the coreclr repo.

**Comment by kangaroo** [comment]
Patches can be consolidated by anyone and everyone, thats kind of the point of git.  See my work with @mikem8361 for reference.  He's leading the LLDB efforts, and when I find issues with his PRs into dotnet/coreclr I submit a PR from kangaroo/coreclr -> mikem8361/coreclr.  He integrates, and the PR reflects that.

**Comment by josteink** [comment]
My gut-reaction was similar to @ajensenwaud and @janhenke, in that I thought having a common workspace/repo/organization (but probably just repo) would be the best way to collaborate on this.

But with that said, I do have some experience with forked repos, and it it does come with some risks and costs, as mentioned by @jkotas and @kangaroo. A very basic example would be that 1. You need to work on your fork, and 2. at the same time ensure you keep up to date with all upstream patches. It doesn't sound like much, but it usually turns out to be a higher workload than you expected, especially when you need to prepare PRs back to mainline.

At the same time I also think it's worth noting the opinions of @kangaroo and similar people who has already contributed to coreclr. If they think a separate organization/repo sounds like overkill, we should respect that opinion as a experience-based one.

So I guess I've changed stance here.

If we go anf try without a seperate "official" BSD-repo, just prepare PRs into official coreclr and this later on proves to be suboptimal for our workflow or progress, surely we can amend that later and create a common repo/org as needed?

**Comment by janhenke** [comment]
IMHO PRs for ongoing development work are just too complicated. Real cooperation in development means I can commit straight away. A PR should only be used to merge the finished feature into the main tree.

I definitely see the point that any fork needs regular updates to stay in sync with the main tree. But that can be done rather easily or even by a bot.

I do think we should have something to make cooperation easy between the members and also something to create a group feeling. Also an easy way to exchange opinion about implementation details which might not be of interest to the general public outside the BSD folks.

**Comment by akoeplinger** [comment]
I agree with @janhenke that it'll be easier to have a separate repo for the BSD work where everyone in the port team can commit than to coordinate PRs between a bunch of people.
That said, if @kangaroo and @jankotas are right and the port will be straight-forward it may not be worth creating it under the dotnet org and hooking up the cla bot etc.

So why not start simple and just give everyone access to @ajensenwaud's repo. It was already mentioned in a couple of places as the BSD bring up fork :smile:

The repo can be abandoned afyer the initial bring up is done and patches get less invasive.

**Comment by josteink** [comment]
Ok. So as far as I can tell, we all agree this should be doable, but we're  just not in agreement about how things should be done.

That disagreement is clearly affecting momentum, and not in a good way. This thread went from lively to dead in about a week.

As far as I can tell, we seem to have two camps:

* One which wants to do their stuff as they see fit and ust submit PRs directly to coreclr as it is.
* Another which wants a non-official repo to work with which tracks mainline (and by proxy group one), where access to commits are easy, and where they can work on partial commits and fixes in a more colleberative fashion, until we have something working.

These two positions aren't really compromisable, but do we have to make that a problem? If we can't agree as a group, do we really need to?

Those who want to submit directly to coreclr can do so, and those who want a seperate repo can work together on that one. Let's just do *something* to keep moving and not kill this issue over something as mundane as this.

If this is the best agreement we can come up with, lets at least agree to that, so that we can move on.

If/when we can get confirmation that everyone is on this, I think it would be time to summon richlander to prepare a repo under the dotnet org, to avoid the "forkers" to flee to far from main base. No offense, @ajensenwaud  :)

So guys... Are we good to go or not? Make some noise.

**Comment by janhenke** [comment]
@josteink I am with you.

@ghuntley What is the current status of the porter box you kindly offered. Are the keys posted here setup to access it?

**Comment by josteink** [comment]
@ghuntley Have you used our github names as user-names? I can't authenticate.

**Comment by ajensenwaud** [comment]
On Mon, Mar 23, 2015 at 6:02 AM, Jostein Kjønigsen <notifications@github.com
> wrote:

> Ok. So as far as I can tell, we all agree this should be doable, but we're
> just not in agreement about how things should be done.
>
> If/when we can get confirmation that everyone is on this, I think it would
> be time to summon richlander to prepare a repo under the dotnet org, to
> avoid the "forkers" to flee to far from main base. No offense,
> @ajensenwaud <https://github.com/ajensenwaud> :)
>
> So guys... Are we good to go or not? Make some noise.
>

Regardless of the approach, I am good to go. :) Even if we just fork the
HEAD from dotnet/coreclr into our own branches and hand around individual
patches, I am happy to go. Let's get the porting started!

**Comment by josteink** [comment]
While we're waiting for @ghuntley...

I do currently have some monthly Azure credit I'm not using for anything, but I didn't find any pre-prepped FreeBSD images available, so I couldn't get anything up there.

Digital Ocean however *does* have some and they have a decent referral program: If you sign up with [this link](https://www.digitalocean.com/?refcode=4a1e84258709) you'll get a $10 credit which is enough to run a 1-core machine with 1GB RAM and 30GB SSD for one month. If you just want a place to work for now until we get something "official" up, you could certainly do worse.

I'm using one of those myself, but with those specs it's obviously not a very shareable machine, hence the suggestion above.

**Comment by ghuntley** [comment]
.... indeed. 

Accounts have been created for the following users:

@janhenke, @josteink, @ajensenwaud & @qbit

```ssh username@portteam.cloudapp.net``

No password, key auth only. All users have the ability to ```su -``` as they belong to ```wheel``.

**Comment by josteink** [comment]
Nice. Never mind my suggestion then. We've got some proper iron now :)

Thanks @ghuntley!

**Comment by qbit** [comment]
can you post the fingerprints for the server?

**Comment by ghuntley** [comment]
Yep, for sure - should have done that in the original post.

@qbit:    

    portteam.cloudapp.net (104.209.35.155) ED25519 key fingerprint
    is 76:84:42:88:49:ee:e5:48:c6:2c:02:c9:fd:9b:8d:da.

@janhenke has been provided the username/password to the Azure Management Portal of this machine and can reboot/shutdown/provision autonomously as needed.

**Comment by josteink** [comment]
If we're otherwise ready to go, I think (for those of use wanting it) setup a seperate collaberative repo to work with. 

@richlander: Can you take care of this?

This repo can then also include a wiki where we share useful stuff, like build/development dependencies, how to cleanly resolve those, etc.

Since @ajensenwaud & me looked at this last time, I can see some new dependencies have been introduced (like libunwind, and lldb for SOS), so I think it makes sense to minimize the effort spent into getting those ready, before we completely mess up ghuntley's Azure VM :)

**Comment by janhenke** [comment]
A short update. I spoke with @richlander on Gitter. FreeBSD will be added to the CI infrastructure once we manage to successfully build coreclr (it does not have to work perfectly, just build cleanly).
I started looking into the next steps for that. See also issue #547.

That should be extra motivation for use to get things going. Come guys, we can make this compile on FreeBSD.

**Comment by janhenke** [comment]
Once #556 gets merged, configuration should work on FreeBSD. So we can now focus on the compiler errors (which start directly with the first file pal.h). Let us do this! We can make it! :+1:

**Comment by josteink** [comment]
Me, @kangaroo  and @ajensenwaud did already resolve most of the compiler-errors in PAL.

I'm trying to find the "correct" way to get a proper version of LLDB compiled for a clean FreeBSD system so that this can get documented (and replicated for a CI-environment).

@janhenke If you already got this handled, would you mind documenting it, and then I can *verify* the documentation by trying to apply it to our Azure VM?

Once that is done, I may start porting PAL-fixes as well.

**Comment by janhenke** [comment]
LLDB is also still an open point for me. I did try to compile it at some point, but in the end skipped that. For the moment it is also not the most important thing, as LLDB is only needed for the SOS LLDB-plugin. The config script outputs just a warning.

I suggest we first try to fix the PAL and get it compiling successfully. Then we can start looking into the finer details of LLDB and the remaining warnings.

**Comment by josteink** [comment]
Just checked and you're right: For now LLDB is just a warning, not a build-breaker so I'll leave that be until later.

As for PAL-fixes, unless you beat me to it, I'll try to see if I can't get (some of) those patches applied to the current coreclr master-branch and see if we can't get any further.

**Comment by emaste** [comment]
I work on LLDB for FreeBSD; the upstream source should build out of the box but will help sort out issues if not. Is there a specific version you need?

**Comment by janhenke** [comment]
CoreCLR currently uses Clang/LLVM 3.5. If you could add LLDB 3.5 to pkg that would be a big help for us.

**Comment by josteink** [comment]
The SOS plugin says either 3.5 or 3.6 will do. I'm looking at the instructions provided [here](https://wiki.freebsd.org/lldb) and that seem simple enough. Only concern I have is that those seem to build from the got master branch and not a specific version, but I can probably get around that.

A PKG source would occlude be even better.

**Comment by janhenke** [comment]
I would like to see it included in pkg for the CI instances setup eventually. Building it on our porter box would certainly work, but for the wider audience I think that is too complicated.

We have to come to a point where we can simply install all build dependencies on a fresh FreeBSD installation by `pkg install bash cmake git libunwind llvm35 lldb35`

**Comment by josteink** [comment]
I've submitted a few PRs for some small issues. One is accepted, the next one is [here](https://github.com/dotnet/coreclr/pull/558).

If that one gets approved, the build will explode in a million new and exciting compiler errors related to registers. The next big job then will be mapping up all the BSD-registers properly to their proper FreeBSD counter-parts.

Once that is done, the build is almost at the point where PAL is fixed and you need to let the build run for quite a while before you encounter new errors. I think this was the point where @ajensenwaud  and I was last time around we tried to get this working.

**Comment by janhenke** [comment]
:+1: 
Happy to this progress!

**Comment by ajensenwaud** [comment]
On Thu, Mar 26, 2015 at 4:42 AM, Jostein Kjønigsen <notifications@github.com
> wrote:

> I've submitted a few PRs for some small issues. One is accepted, the next
> one is here <https://github.com/dotnet/coreclr/pull/558>.
>
> If that one gets approved, the build will explode in a million new and
> exciting compiler errors related to registers. The next big job then will
> be mapping up all the BSD-registers properly to their proper FreeBSD
> counter-parts.
>
> Once that is done, the build is almost at the point where PAL is fixed and
> you need to let the build run for quite a while before you encounter new
> errors. I think this was the point where @ajensenwaud
> <https://github.com/ajensenwaud> and I was last time around we tried to
> get this working.
>

We didn't get PAL ported in its entirety, but we definitely made good
progress. I will start walking through PAL now, starting with the build
scripts.

Cheers

Anders

**Comment by josteink** [comment]
Another two PAL patches going in:

https://github.com/dotnet/coreclr/pull/573
https://github.com/dotnet/coreclr/pull/573

This should leave us the runlocale-thing (which I don't remember the resolution for. clang35?), and the MC/BSD-register entries.

**Comment by josteink** [comment]
After giving it a go, I must admit it feels nice being able to submit FreeBSD-patches, and having the Windows and Linux CI-builds go at the same time.

I've once accidentally broken the Linux-build, and having it happen on every small commit, rather than on one big giga-commit with "all" our patches makes it much easier to fix up. Basically working on the main repo makes me feel much more confident in the patches submitted.

So I'm definitely a convertee. I don't see the need for a coreclr-freebsd repo. If anyone else really wants it I'm not against that, but they'll have to shout up to richlander themselves ;)

**Comment by josteink** [comment]
All my PRs so far are accepted and good, which leaves us the tricky job getting PAL's context.cpp right.

I'm short on time, but last time around @kangaroo was really helpful, and tried to come up with something working. That's available in his repo here in the freebsd-branch:

https://github.com/kangaroo/coreclr/tree/freebsd

Definitely worth looking into for those who have time.

**Comment by josteink** [comment]
Another observation. The build output seems to hint that we're in a scenario were ASSIGN_REG is not getting defined:

    [  0%] Building CXX object src/pal/src/CMakeFiles/CoreClrPal.dir/arch/i386/context.cpp.o
    /home/josteink/build/coreclr/src/pal/src/arch/i386/context.cpp:316:9: error: use of undeclared identifier 'Rbp'
	    ASSIGN_ALL_REGS
	    ^
    /home/josteink/build/coreclr/src/pal/src/arch/i386/context.cpp:234:9: note: expanded from macro 'ASSIGN_ALL_REGS'
	    ASSIGN_CONTROL_REGS \
	    ^
    /home/josteink/build/coreclr/src/pal/src/arch/i386/context.cpp:192:20: note: expanded from macro 'ASSIGN_CONTROL_REGS'
	    ASSIGN_REG(Rbp)     \
		       ^
    /home/josteink/build/coreclr/src/pal/src/arch/i386/context.cpp:469:13: error: use of undeclared identifier 'ptrace_registers'
		ASSIGN_CONTROL_REGS
		^


The first register to fail out is Rbp. If we check the source at line where it all gets invoked (469), we see this code preceeding it:

https://github.com/dotnet/coreclr/blob/master/src/pal/src/arch/i386/context.cpp#L462-L466
````c++
#if HAVE_PT_REGS
#define ASSIGN_REG(reg) PTREG_##reg(ptrace_registers) = lpContext->reg;
#elif HAVE_BSD_REGS_T
#define ASSIGN_REG(reg) BSDREG_##reg(ptrace_registers) = lpContext->reg;
#endif
````

Notice how that is an "if, else if" but no default "else" condition to handle the rest.

According to the configuration process we don't have PT_REGS and we don't have BSD_REGS either. Same with GREGSET_T.

From CMakeCache.txt:

    //Result of TRY_COMPILE
    HAVE_PT_REGS:INTERNAL=FALSE
    //Result of TRY_COMPILE
    HAVE_BSD_REGS_T:INTERNAL=FALSE
    //Test HAVE_GREGSET_T
    HAVE_GREGSET_T:INTERNAL=

Is this the error, or is there something else obvious which I'm missing? What regs are we supposed to have besides BSD-regs?

I recall in @ajensenwaud's repo, we just manually configured for FreeBSD and then this wasn't a problem. Did we do it wrong or is the auto-config wrong?

Until we get this solved, the build wont move forward.

**Comment by kangaroo** [comment]
@josteink Check the cmake output to see why BSD_REGS_T isn't being set?

**Comment by josteink** [comment]
Now that would be too easy, now wouldn't it? Sent a PR to get it fixed. :)

https://github.com/dotnet/coreclr/pull/584

**Comment by josteink** [comment]
Ok. So I've gotten most, but not all of the BSD-registers fixed.

Unfortunately I'll be "off" for the Easter and wont be around to hack on this material. Anyone else feel like taking over for now to keep the momentum up? @janhenke ? @ajensenwaud ?

The registers remaining weren't around last time we tried to get this thing ported, [basically just the FPREGs](https://github.com/dotnet/coreclr/blob/master/src/pal/src/arch/i386/context.cpp#L78-L90), so I didn't have any quick-fixes to transfer.

As far as I can tell this is defined in `/usr/include/sys/ucontext.h`. We just need to find a proper trap for our `#define` entries, if there isn't already a good one. Or do we? If we do, see [my earlier commit here](https://github.com/dotnet/coreclr/commit/869e47aa7a0b807fac2076f3bba33e50bfc248f9) for an example.

What I'm also interested in is how are our FreeBSD-fixes so far playing out with the rest of the BSD-verse? Have things improved? Are the Darwin and OpenBSD builds also benefiting from these commits? Or are we creating a conflict of sorts?

I would love to get some feedback on this. @kangaroo ? @qbit ?

Anyhow to everyone: Happy easter!

**Comment by josteink** [comment]
Created seperate issue to discuss FPREGs, with some preliminary code.

https://github.com/dotnet/coreclr/issues/594

**Comment by ajensenwaud** [comment]
Well done, Jostein.

I will have a look at the FPREGs now, I don't know yet how we can tackle
them.

For other *BSDs I will spin up VMs up now to see how well it compiles on
NetBSD and OpenBSD (amd64). I will see if I can find some time to test it
on Darwin as well.

Cheers

Anders

On Sat, Mar 28, 2015 at 9:56 PM, Jostein Kjønigsen <notifications@github.com

**Comment by josteink** [comment]
Sounds great! 

When you first have VMs up and running, I would love it if you could also verify my preliminary work on fpregs other places than freebsd.

Would be neat if we didn't have to do this messy register-work more than once :)

**Comment by phillipkent** [comment]
Hello @janhenke and others...

A number of developers here at http://www.interoute.com in London are interested in FreeBSD and your .NET port project.

Our company would like to help out by offering free use of virtual machines in our cloud platform, Interoute VDC [http://cloudstore.interoute.com/main/WhatInterouteVDC]. 

We have a FreeBSD template already installed, and you can upload other templates as you require them (you need an OVA based on VMware vmdk image, or an ISO source).

Can one of you get in touch with me directly [phillip.kent@interoute.com] and we can discuss how much resources to set up....

Regards from London.
Phillip

**Comment by ajensenwaud** [comment]
OK, I have been working on setting up a NetBSD VM -- more work is required
on my part to adjust the build scripts to NetBSD first as NetBSD/amd64 is
not picked up by build.sh. Will let you know once I have progressed with
this.

On Mon, Mar 30, 2015 at 5:27 PM, Jostein Kjønigsen <notifications@github.com

**Comment by josteink** [comment]
@phillipkent Thanks for the offer. Right now @ghuntley has our back covered as far as FreeBSD-instances go. We've also been offered Azure resources from @MattWhilden at Microsoft.

OpenBSD and NetBSD is currently not able to run on Azure, so if you could host VMs for those instances, that might be useful.

@ajensenwaud : Once "done" with your machines, do you think it would be possible to have them uploaded to interroute? That way others could help out testing/porting.

**Comment by josteink** [comment]
Out of curiousity, I decided to cheat on the FPU registers, and just commented out the lines involving the remaing two (XMM etc). Basically just tosee what the next hurdle in the build is.

Disappointingly so, I discovered that just like the FPU registers, other parts of coreclr is progressing faster than we're able to port it. PAL's [context2.S](https://github.com/dotnet/coreclr/blob/master/src/pal/src/arch/i386/context2.S) (mostly assembly + macros) which used to build is now complaining about basic assembly instructions being invalid.

I suspect some of the macros defined might be messing up evaluation on FreeBSD. Any assembly guys around wanting to take a look? Should we create a seperate issue for this as well?

**Comment by phillipkent** [comment]
@josteink Great. I will create a user account, and my colleague @xied75 is looking into running OpenBSD and NetBSD VMs on our platform.

**Comment by josteink** [comment]
I made a separate issue for the assembly failing the build. I think itmakes sense to allow discussion on a per bug basis, and it probably makes iyt easier to track if we're getting anywhere and at what speed.

Assembly issue: https://github.com/dotnet/coreclr/issues/615

**Comment by MattWhilden** [comment]
Highly recommend breaking these into separate issues. Smaller increments helps two ways:
1) We can better understand the work that's outstanding.
2) It's easier to point to the work already done and get the warm and fuzzy feels.

Joking aside, it's really nice internally to MS to be able to easily articulate the work that y'all have undertaken. #everyonehasaboss #yallareawesome

**Comment by josteink** [comment]
I agree with @MattWhilden and have created new issues ongoing as I've made further progress.

@janhenke This issue seems to have managed to summon a team of sorts, and gotten ourselves bootstrapped. And as such this issue has achieved its mission, right? I think it's time to close it at this point. Agreed?

**Comment by janhenke** [comment]
@josteink For the initial mission I agree. Before it is closed though, I want to remark for any potential future reader that this is by no means a closed group. Any volunteer is welcome to continue making CoreCLR run smoothly on FreeBSD as well as the other BSDs.

**Comment by MattWhilden** [comment]
Yep. I'll try to keep things tagged appropriately so that new folks can know where to help out. Great start!

**Comment by ajensenwaud** [comment]
Hi Jan,

I am getting a hard error when trying to compile on OpenBSD because
libunwind is not available (and not available from the ports tree either).

Commencing CoreCLR Repo build
Setting up directories for build
Checking pre-requisites...
Commencing build of native components for OpenBSD.x64.Debug
Invoking cmake with arguments: "/home/aj/coreclr" DEBUG
which: clang++-3.5: Command not found.
CMake Warning at src/ToolBox/SOS/lldbplugin/CMakeLists.txt:26 (message):
  Cannot find lldb-3.5 or lldb-3.6.  Try installing lldb-3.6-dev (or the
  appropriate package for your platform)


-- Looking for include file libunwind.h
-- Looking for include file libunwind.h - not found
CMake Error at src/pal/src/configure.cmake:886 (message):
  Cannot find libunwind.  Try installing libunwind8 and libunwind8-dev (or
  the appropriate packages for your platform)
Call Stack (most recent call first):
  src/pal/src/CMakeLists.txt:3 (include)


-- Configuring incomplete, errors occurred!
See also
"/home/aj/coreclr/bin/obj/OpenBSD.x64.Debug/CMakeFiles/CMakeOutput.log".
See also
"/home/aj/coreclr/bin/obj/OpenBSD.x64.Debug/CMakeFiles/CMakeError.log".
Failed to generate native component build project!

**Comment by ajensenwaud** [comment]
@qbit, I have set up an account for your on our new OpenBSD instance in
Hong Kong:

IP: 103.232.35.201 (port 2201)
User: qbit

With the SSH key above.

Others -- please let me know if you would like access as well.

cheers

Anders

**Comment by janhenke** [comment]
I think the first step for OpenBSD is to have Clang 3.5 (or 3.6) and libunwind. libunwind is necessary for the exception handling afaik (stack unwinding to find the exception handler).

How is the current status of LLVM/Clang on OpenBSD?

**Comment by ajensenwaud** [comment]
LLVM/Clang can be installed as a binary package or from the Ports
collection.

**Comment by janhenke** [comment]
Which versions? CoreCLR currently compiles with 3.5. Is lldb available too? If yes we would need lldb 3.6.

**Comment by ajensenwaud** [comment]
$ clang -v
clang version 3.5 (trunk)
Target: amd64-unknown-openbsd5.6
Thread model: posix

lldb is not in ports, but I will see if it builds manually.

**Comment by josteink** [comment]
Looks like openbsd work is progressing nicely. Which is good news indeed.

But I think it would be better if you created a separate issue for this, like "openbsd build-machine must be configured" or something similar. 

Sounds OK?

**Comment by qbit** [comment]
Thanks @ajensenwaud ! I seem to be unable to access said machine.

> ssh: connect to host 103.232.35.201 port 2201: Connection timed ou

**Comment by ajensenwaud** [comment]
Hi Aaron, not sure what the issue was -- can you please try again, this
time just on port 22? I added an A record for the host so you can access it
via freebsd-hk.zbsd.org (yes, we have to change that to something more
descriptive, but it will take you to the OpenBSD machine).

**Comment by RussellHaley** [comment]
Hi There, 

Just starting to play with the new drop of Core 1.0. I don't see a lot of FreeBSD support in the documentation other than a port of the CLR, but maybe that's not significant as once the CLR is built the rest is managed code?

**Comment by josteink** [comment]
It's not as simple as we'd like it to be.

Initially coreclr (and corefx) on non-Windows platforms was mostly bootstrapped using mono.

So when we did the original FreeBSD port (and I guess we got to around 98% or something like that, just a few platform-specific bits of corefx missing), we could basically leverage mono which was already available and working.

At that point we kind of lost momentum for a variety of reasons and that's a shame. But .NET itself kept on moving (adding awesome things like the Dotnet CLI which is a massive improvement).

Part of that process involved the .NET system to be rewritten to be self-bootstrapped. And with that it drew in lots of common .NET ecosystem dependencies: nuget, roslyn, etc.

This has pros and cons. The pros are obvious: We no longer depend on mono to build .NET.

The cons is that to build .NET Core from source, smoothy, without too much nonsense, you now need to download a prebuilt .NET Core SDK (including its own functional coreclr runner) to run the .NET toolchain used to build .NET.

**And that .NET toolchain is only available for *already fully supported* platforms (which does not include FreeBSD).**

Building this SDK chain yourself can be done, but it's not trivial. Some people are working on it here: https://github.com/dotnet/corefx/issues/10610

Others are trying to piece together what we have to create a proper FreeBSD ports package: https://github.com/dotnet/coreclr/issues/6115

From those 2 issues, you should be able to find references to other issues and leads which may get you the whole way through. But that's not exactly a as smooth ride as anyone should have to suffer.

Basically this whole self-bootstrapping for everything good it does, has (for now) inadvertently set back a working FreeBSD-ports quite a bit. 

So that's the bad news. The good news is that any PR which improves the situation quickly gets accepted.

If someone wants to pick up all the code we ported, and help finally land the thing, I'm sure everyone is going to be grateful about that.  I'll be more than willing to assist with code-review and other non-intensive stuff.

**Comment by RussellHaley** [comment]
Thanks @josteink that brings me up to speed. I'll review everything carefully. As a curiosity, would it be possible to go back to bootstrapping with mono (no need to get too detailed in your answer)?

**Comment by josteink** [comment]
It might be. It could definitely be worth testing and trying, although you'll probably have to write mono wrapper-scripts for the entire SDK tool chain (Dotnet CLI, nuget, Roslyn, msbuild, etc).

In a sense that would be creating a new SDK platform package, except the platform is mono, not a specific OS. That would probably be useful for all the platforms which aren't already bootstrapped.

If that could be made, that might make an interesting fallback on currently "unsupported OSes".

If that's more work or less than just getting a FreeBSD SDK building, I don't know. It would be starting from scratch, as apart from finishing something already started.

On the flip side, you would probably be able to get going at it without having to wait for anyone else to do work for you.

Whatever you decide to do, I hope it works out :)

**Comment by RussellHaley** [comment]
Ha, I just got Monodevelop 6.1 going on freebsd again. :P

I need to get my head around the build toolchain first so let me go back over all the comments...
